### ISSUE TITLE
Unable to find address mapped as node with addr:city & addr:street & addr:housenumber

### ISSUE THREAD

**Issue by bentolor** [organicmaps/organicmaps#3307]
Thanks for this awesome piece of FOSS software which has become one of my number 1 tools on my android phone.

## Issue
I'm unable to find a location "Karlstraße 3" which is mapped as node with `addr:city`, `addr:street`, `addr:housenumber`, `add:postcode` and `addr:country`.

## Steps to reproduce
1. Navigate to [48.4042014, 9.9860426](https://www.openstreetmap.org/#map=18/48.4042014/9.9860426)
2. Search for "Karlstraße 3" or "Karlstr. 3"

## Result
OM shows first 1:1 results in 25km distance

## Expected result
Find the actual result on spot which is OSM node [#2293529605](https://www.openstreetmap.org/node/2293529605)

Tag | Value
----|-----
[addr:city](https://wiki.openstreetmap.org/wiki/DE:Key:addr:city?uselang=de) |	Ulm
[addr:country](https://wiki.openstreetmap.org/wiki/DE:Key:addr:country?uselang=de)| 	DE
[addr:housenumber](https://wiki.openstreetmap.org/wiki/DE:Key:addr:housenumber?uselang=de)| 	3
[addr:postcode](https://wiki.openstreetmap.org/wiki/DE:Key:addr:postcode?uselang=de) |	89073
[addr:street](https://wiki.openstreetmap.org/wiki/DE:Key:addr:street?uselang=de) |	Karlstraße

**Comment by TheAdventurer64** [comment]
Try search up Karlstraße 3 and see if anything pops up.

**Comment by biodranik** [comment]
It works on my iOS. Did you try the latest app version with the latest map data?

**Comment by bentolor** [comment]
@biodranik Yes!  I'm on latest F-Droid version 2022.08.23-5-FDroid with up-to-date mapdata.

**Comment by bentolor** [comment]
Thanks for testing and your reply, @biodranik! 

This is…weird. The behaviour is most of the time like described above. But after **several screen recording runs**, the **expected match appeares as well**.

## My observations
* Search takes very, very long. The initial run takes ages. Look here. The search starts at 00:21 and results are at 01:00 (missing the expected result). The actual hits are in a nearby city: ![Early run, no hits](https://user-images.githubusercontent.com/588260/188139792-260864a5-2b75-4823-bd99-7d3195c63a38.mp4)
* Duration decreases after several runs: ![intermediate-run-no-hit](https://user-images.githubusercontent.com/588260/188139943-ca768b04-20da-4996-91da-d8bab4b91a72.mp4). Still first hit is in _Niederstotzingen_, a village nearby.
* After *several runs* I ![do receive a direct hit!](https://user-images.githubusercontent.com/588260/188140734-d70e0c7a-ff42-464b-b4d9-0be60985f80b.mp4). Note the pause – that's because I was stumbled over the actual hit!


## My guess
I assume this is some sort of time-cap which kicks in here and truncates results. **Probably your best & overall cheapest answer is: Buy a new phone!**

My Phone is a [OnePlus 3T still having rather current spects](https://en.wikipedia.org/wiki/OnePlus_3T) like Qualcomm Snapdragon 821 Quadcore, 6GB RAM – but doe to it's age I suspect the Disk I/O degraded.

1. Does this make sense? 
2. Which phone do you recommend? ;-)

**Comment by biodranik** [comment]
Thanks for the videos, that's even more strange ) The current search timeout is... 8 seconds, that is definitely not your case for the first video. @vng can you explain it?

@vng anyway, the search results from the viewport should be _in the top from the start_, it's where a user is looking and expects them in most cases.

Regarding the slow phone... That is an excellent case to make our search faster somehow ) It would be great to profile our code on devices like yours ) Maybe the memory has degraded over 8 years of extensive usage, who knows. Any newer device should be faster.

**Comment by vng** [comment]
@biodranik 8 seconds timeout is kinda _soft_ deadline, which doesn't mean immediate search stop

@bentolor Am I right that you have turned-off you current position (probably first result is closer to your location or your last known location if any ..)? In that case ranking looks weird here, I agree. Could you please write us the full list (screenshot) of downloaded maps to check/test speed issues. But I suspect that "Karlstraße" is quite common street name and I can reproduce it on any German maps set ;)

**Comment by bentolor** [comment]
@vng I only disabled location for the screenrecording, so it won't reveal my exact location. Behaviour is the same; and the "25km distance" mentioned in the ticket is the closes hit of the search results vs. the actual 3~5km Karlstadt nearby. BTW: The ranking/subset of result was mostly consistent regarding first hit.

Not sure if @biodranik assumes it's just the list order: I scrolled through the results and in fact it seems the nearby hit is missing at all in those cases. That's why I assumed a truncation of search results due to a time-cap.

My downloaded set of maps is rather limited, though:  _Baden-Württemberg/Regierungsbezirk Stuttgart, Regierungsbezirk Tübingen, Bayern/Oberbayern/München, Schwaben_ plus _World_ and _World Coast_

@biodranik Gosh – am I _that_ old? 5,5 years old intensive usage. Got it 12/2016. To verify my assumption I did a disk speed test using CPDT. See screenshot below. So sequential r/w seems got, random r/w sub-par.

![image](https://user-images.githubusercontent.com/588260/188261670-1941e732-af22-46e8-99f8-b732fca76144.png)
![image](https://user-images.githubusercontent.com/588260/188261673-192ab0f6-9136-4ab3-8833-4f72bb14269b.png)

**Comment by biodranik** [comment]
I think here the slow speed is related to both: our algorithms that scan a lot of data in files, and slower read speed (compared older iPhones, for example, have 5x-10x faster read speed).

@vng I reproduced the same behavior with several german maps downloaded. When the search is cut by the timeout, there is no that result even if it's in the viewport's center. The next launch (features/indexes are already cached), properly shows it.

**Comment by vng** [comment]
Investigate search speed with this example.

**Comment by vng** [comment]
OMG, there is also Neu-Ulm town with the same address: https://www.openstreetmap.org/node/768543234#map=19/48.39355/10.00103

But OM show that this is Ulm.

**Comment by vng** [comment]
Hm, search on my iPhoneX with _all_ Germany downloaded takes about 10 seconds. But with my current fixes ..

**Comment by bentolor** [comment]
Thanks Viktor @vng for taking a look into that. I've already seen in the merge request that you built a test case on that scenario. Really appreciate the efforts. 

I think I also stumbled upon that Neu-Ulm match but assumed a wrong mapping the source data. I'm really sorry that this issue turns out to be that _"I'll fix that in a few minutes"_ to "…another bug here?".

Currently in comparison to Google Maps i..e the search could be one of the major area, where improvements will add real value for the users. That's why I tried to enhance what I could do #3309. Currently I'm regularly switching to Google maps just to find the location and manually copying the result back to OM for greater detail.

Btw: Whats the intention of your fixed? I'm not fluid in C++. Is it to search in the local zone first so that nearby hits have a better chance to show up? 
Again, thank you very much!
